src/collapse/collapse-content.tsx

```diff
@@ -0,0 +1,127 @@
+import React, {
+  useState,
+  useEffect,
+  useRef,
+  useContext,
+  useMemo,
+  useCallback,
+  PropsWithChildren
+} from 'react';
+import classNames from 'classnames';
+
+import dataTests from '../global/data-tests';
+
+import {getElementHeight, toPx} from './utils';
+import CollapseContext from './collapse-context';
+import {COLLAPSE_CONTENT_TEST_ID, COLLAPSE_CONTENT_CONTAINER_TEST_ID} from './consts';
+
+import styles from './collapse.css';
+
+const DURATION_FACTOR = 0.5;
+const DEFAULT_HEIGHT = 0;
+const VISIBLE = 1;
+const HIDDEN = 0;
+
+type Props = {
+  minHeight?: number;
+  className?: string;
+  'data-test'?: string | null | undefined;
+  disableAnimation?: boolean;
+};
+
+/**
+ * @name CollapseContent
+ */
+
+export const CollapseContent: React.FC<PropsWithChildren<Props>> = ({
+  children,
+  minHeight = DEFAULT_HEIGHT,
+  disableAnimation = false,
+  'data-test': dataTest
+}) => {
+  const {collapsed, duration, onChange, id} = useContext(CollapseContext);
+  const containerRef = useRef<HTMLDivElement>(null);
+  const contentRef = useRef<HTMLDivElement | null>(null);
+  const initialContentHeight = useRef<number>(minHeight);
+  const contentHeight = useRef<number>(DEFAULT_HEIGHT);
+  const [dimensions, setDimensions] = useState({
+    width: 0,
+    height: 0
+  });
+  const [height, setHeight] = useState<string>(toPx(minHeight));
+  const [showFade, setShowFade] = useState<boolean>(true);
+
+  const onTransitionStart = useCallback(() => {
+    if (!collapsed) {
+      setShowFade(false);
+    } else {
+      setShowFade(true);
+    }
+    onChange(collapsed);
+  }, [collapsed, onChange]);
+
+  useEffect(() => {
+    const ref = containerRef.current;
+    if (ref) {
+      ref.addEventListener('transitionstart', onTransitionStart);
```

Comment:
Wouldn't this event be triggered by bubbled event if some user content has been animated inside <CollapseContent/>?

Reply:
@andrey-skl  Yes, you're right. To be honest, this addEventListener is redundant and can be removed.